### ISSUE TITLE
Allow using dart:mirrors

### ISSUE THREAD

**Issue by stevenroose** [flutter/flutter#1150]
I know that mirrors is blocked now because it causes the generated packages to be very large.

However, on the long term, it should remain a goal to find a solution for this. Many libraries use mirrors so not being able to use mirrors for Flutter apps will make the framework a lot less powerful.

The trade-off between large package sizes and using mirrors should be left to the user. A possible solution might be to use an explicit `--allow-mirrors` flag. But of course the ultimate solution should be to find a way to use mirrors without blowing up package sizes.

Java has reflection and Objective-C has Runtime Reference, so why would Flutter not have a reflection system?

**Comment by Hixie** [comment]
Allowing mirrors would mean we couldn't remove unused code (since all the code is implicitly used). That's what causes the code size increase.

Java apps are huge, that's one of the things we want to avoid. Objective C, as I understand it, doesn't trim unused members from classes either. It also doesn't strip any classes that are exported, because it uses dynamic linking.

Dart uses static linking (effectively), and using static analysis we can strip out anything that isn't used ("tree shaking"). If you import a huge Dart library but only use a self-contained two-line method, then you only pay the cost of the two-line method, even if that Dart library itself imports dozens and dozens of other libraries.

(At least, in theory. I'm not sure how much of this we implement yet.)

We could enable mirrors only for classes that you opt into (similar to Delphi's `$M+`). That would require support from the language.

**Comment by stevenroose** [comment]
@Hixie Isn't that what the [`@MirrorsUsed`](https://api.dartlang.org/135611/dart-mirrors/MirrorsUsed-class.html) annotation is for? I thing dart2js also does tree shaking using the `@MirrorsUsed` annotation.

**Comment by Hixie** [comment]
I think that's more about avoiding name mangling, which is a whole other kettle of fish (we don't do any name mangling). But it would be a similar kind of thing, yes. To my knowledge the snapshotter doesn't support any annotation of that kind currently.

**Comment by stevenroose** [comment]
I quote from [this article](http://blog.sethladd.com/2014/01/how-to-shrink-size-of-your-dart-app.html) from @sethladd 

> Mirrors are really powerful, but they introduce challenges to dart2js's tree shaking abilities. Normally, dart2js can look at the entire app and deduce what code is required to run the app. However, as of the time of this writing, mirrors (due to their API design and highly dynamic nature) more-or-less disable dart2js's tree shaking.
> 
> Enter @MirrorsUsed, a metadata annotation, which explicitly states what parts of the program are reflected. Tools like dart2js can use @MirrorsUsed to re-enable tree shaking.

**Comment by Hixie** [comment]
Ah, yeah, I didn't see 'targets' and 'metaTargets' before. Those could be used to do this, if the snapshotter supported them.

**Comment by sethladd** [comment]
MirrorsUsed is not supported and has edge cases it doesn't cover. I wouldn't build a solution around it.

**Comment by stevenroose** [comment]
@sethladd I noticed it said "experimental", yes.
But is it still the only thing that helps dart2js tree shaking with mirrors? Or are there other strategies.

Anyways, dart2js and Flutter seem to have a similar issue when it comes to mirrors usage, so I guess they can use the same "solution", however complete that solution might be.

I believe dart2js still warns users of large output size when it sees mirrors usage, that is a good practice. Better than preventing users from using it...

**Comment by floitschG** [comment]
There is also a reflectable package. It can either run untransformed, using
dart:mirrors, or transformed, where it extracts the information beforehand.
It is not complete, though. For example there is no way to access private
identifiers.
It also adds an additional build step...

**Comment by eernstg** [comment]
For completeness of reflectable, please check out the 'Known limitations' section at the end of https://github.com/dart-lang/reflectable/tree/master/reflectable (noting that the methods `delegate`, `isAssignableTo`, and `isSubtypeOf` have been implemented and will be in the next published version).

**Comment by abarth** [comment]
We seem to have gotten pretty far without dart:mirrors.  I think we can safely close this issue.

**Comment by stevenroose** [comment]
I don't agree.

> We seem to have gotten pretty far without dart:mirrors.

You know that it makes no sense to say that? People know that mirrors is not supported so they either leave the platform (I don't think you can deny that traction is not really great), removed features or used alternative libraries.

Both Android and iOS support reflection, which means that as a competitor against their default development infrastructure, this will always be a setback for some developers.

While reflectable is a decent enough way to bridge the gap for now, it is far from ideal and as I already pointed out, dart:mirrors + MirrorsUsed should be able to get exactly the same performance than reflectable.

**Comment by abarth** [comment]
The approach we use when considering which features to implement is to compare the use cases for those features against the cost of implementing and maintaining the feature.  In the case of `dart:mirrors`, we've largely found less costly ways of addressing these use cases.  It's possible that we'll eventually encounter a compelling use case that requires `dart:mirrors`, in which case we can reconsider implementing that feature, but having this bug open doesn't help us make progress.

Rather than bugs requesting particular solutions, we prefer to work from bugs that state problems, which gives us the flexibility to address those problems in a different ways and potentially find better solutions.

**Comment by stevenroose** [comment]
Well, I agree that this is not of a particular priority, that's for sure.
But is there a problem withs issues that live a long time? You can tag them
as improvement or postponed or low priority. I just think closing an issue
that still is one just because you want to reduce the backlog, is a
strategy I don't really agree with.

**Comment by Hixie** [comment]
The problem is this isn't really an issue. It's not describing something you can't do, it's just a request for a particular solution. We prefer to track the actual issues. You should definitely feel free to file bugs describing the things you want to do that you can't do without mirrors.

**Comment by stevenroose** [comment]
@Hixie To quickly come back to this: the concrete issue it not about a specific functionality that is missing, but more the inability to use any library that relies on mirrors.

I think a lot of libraries will have to start doing things like this: https://github.com/PointyCastle/pointycastle/issues/107

**Comment by Hixie** [comment]
Those libraries are going to stop working in general once Dart drops mirrors entirely, so I'm not too worried about that.

There are certain things that aren't available to a Dart library in Flutter. Some of those things are obvious and nobody thinks about them, like "C++ syntax" or "Perl libraries". Others are less obvious initially, like "no dart:js" or "no dart:html", but they make sense with a quick examination. And then there's some that aren't obvious at all unless you really know how it works, like "no dart:mirrors". There's libraries affected by all of these. I don't think it really matters on the long run. People will just avoid the libraries that don't work with Flutter because of using dart:mirrors just like they avoid using the libraries that don't work with Flutter because they're written in Lua, or whatever.

**Comment by stevenroose** [comment]
Are there plans to discontinue mirrors entirely?

I completely disagree with your reasoning btw. For the things you 
mention there are valid reasons that *prevent* these features being 
available in Flutter: Dart does not support Perl syntax or Flutter apps 
have no notion of a DOM. There is nothing that *prevents* mirrors from 
being available in Flutter. The reason it is not supported is binary 
size, which is exactly the same concern for dart2js. In dart2js however 
this trade-off is left to the user.

Don't get me wrong, I totally understand that support for mirrors is a 
lot of work and there might be valid arguments that it's not worth it. 
However, it is a feature that people may want and it is possible to 
support, so the place for it should be on the long-term wish list.


On 4 Nov 2016 20:29, "Ian Hickson" <notifications@github.com 
<mailto:notifications@github.com>> wrote:

    Those libraries are going to stop working in general once Dart drops
    mirrors entirely, so I'm not too worried about that.

    There are certain things that aren't available to a Dart library in
    Flutter. Some of those things are obvious and nobody thinks about
    them, like "C++ syntax" or "Perl libraries". Others are less obvious
    initially, like "no dart:js" or "no dart:html", but they make sense
    with a quick examination. And then there's some that aren't obvious
    at all unless you really know how it works, like "no dart:mirrors".
    There's libraries affected by all of these. I don't think it really
    matters on the long run. People will just avoid the libraries that
    don't work with Flutter because of using dart:mirrors just like they
    avoid using the libraries that don't work with Flutter because
    they're written in Lua, or whatever.

    —
    You are receiving this because you authored the thread.
    Reply to this email directly, view it on GitHub
    <https://github.com/flutter/flutter/issues/1150#issuecomment-258526243>,
    or mute the thread

**Comment by Hixie** [comment]
Nothing prevents us from having a Perl interpreter either. :-)

The question is, what can't you do with Flutter today that you _could_ do if you had mirrors? "Use libraries that use mirrors" is not a good answer to that, because it has the simple retort "just rewrite the libraries not to use mirrors", just like "You should support Perl libraries so that I can use Perl libraries" isn't a good description of a problem (just write those libraries in Dart instead).

What actual thing can't you do today that you could do with mirrors? You should file a bug saying that you can't do that specific thing. If mirrors are the best way to address that problem, and if the problem is worth addressing, then we'd figure out how to introduce mirrors.

> Are there plans to discontinue mirrors entirely?

I don't know about specific plans but with the push towards strong mode it would make sense to me to drop things like mirrors. (Personally I'd like to replace mirrors with metaclasses, so we get all the benefits of mirrors with none of the tree-shaking costs, but that's also probably not going to happen.)

**Comment by eernstg** [comment]
> Are there plans to discontinue mirrors entirely?

Drive-by comment: This is the Flutter repo, so this discussion is distinct from discussions about how/whether Dart will support reflection on other platforms. Besides, as long as it's possible to edit Dart code manually it will also be possible to use techniques like the ones in reflectable to obtain some level of support for reflection. In that sense it isn't even possible to discontinue mirrors entirely.

**Comment by branflake2267** [comment]
Adding a note for future searches. I can't use the json_object library, it depends on dart_mirrors, which is blocked. Would be nifty if the IDE would detect blacklisted libraries or transitives that are blocked. 

pubsec.yaml "json_object: any"
import 'package:json_object/json_object.dart';

**Comment by xster** [comment]
Semi tangential: seems like json_object does 2 things: 

1. Access json map like an object
2. Serialize dart objects into json

I assume use case 1 is the most popular and it only has a mirrors dependency from this [line](https://github.com/chrisbu/dartwatch-JsonObject/blame/master/lib/json_object.dart#L220) which is 5 years old. It should be still doable without mirrors [like mockito](https://github.com/dart-lang/mockito/blob/master/lib/src/mock.dart#L479)

**Comment by zoechi** [comment]
@branflake2267 I think pub.dartlang.org is taking this into account now and doesn't show the `flutter` tag when `mirrors` is used.
https://pub.dartlang.org/packages/json_object

I think currently 
https://pub.dartlang.org/packages/json_serializable
https://pub.dartlang.org/packages/built_value
are the best options for JSON serialization
besides several others (jaguar, dson, ...)

**Comment by CrazyStaff** [comment]
I think this might be a good case to support mirrors, so that Json could be easily 
serialized to objects.

https://github.com/redstone-dart/redstone_mapper

**Comment by matejthetree** [comment]
Also for dependecy injection like in this package.
In big apps with a lot of business logic, this is a dream.
https://github.com/dotdotcommadot/dart_robotlegs_di

**Comment by asxhl2007** [comment]
@stevenroose I will leave this platform, really!😂

**Comment by kidcosmic** [comment]
Also, ORM data majicks.

IO is inherently dynamic and untyped and we live with that, but in the case of the M that drives our V, we shove all the type sickness of data conversion into an automated small place, so we can thoroughly enjoy type safe models in the rest of our world. Model is so inanely crucial, pervasive, and mutating in our many, highly iterative applications, that automated, reflecting, self inflating, IO untethered data models, is something we thrive on.

Also, you are all aware of this, yes?

### Flutter - Beautiful native apps

# in record time

---

_Bring dart:mirrors_

**Comment by kidcosmic** [comment]
To further iterate the point, you absolutely cannot beat:

```
class User extends GGDBObject {
    String id, username, avatar;
    List<User> friends;
}
User one = User.get("1");

var two = User();
two.username = "kidcosmic";
two.friends.add(one);
two.save();
GGNetworkLibrary.registerUser(two);
```

A class definition being the only thing necessary to leverage both DB IO and network IO for models is _record time_.

`@MirrorsUsed` is far more than enough to satisfy tree shaking. Literally saying don't shake my User class is obviously and apparently sufficient. Mirrors are widely and by far mainly used for data classes which are never shaken anyway. Instead of saying no to people that want to reflect data classes, say no (or if ever sufficient bandwidth) to people that want to reflect x for some random edge case reason.

Also, I agree that tree shaking is a phenomenal optimization and worth protecting, but I do believe that mirroring is achievable without diminishing from tree shaking at all, particularly in the case of data classes. It's apparent that a `@MirrorsUsed` implementation that simply excludes a class's own defined properties/constructors from shaking is sufficient. Mind you, it's most likely that only a majicked data class's constructors are candidates for shaking and only because dart classes don't inherit constructors.

**Comment by jinseoung-lee** [comment]
Any news? Really??? 😢😢

**Comment by zoechi** [comment]
@jinseoung-lee any news about what?
It was a deliberate decision to not include `dart:mirrors` for concrete reasons.
There are alternatives listed above like code custom generation or the reflectable package.

**Comment by jinseoung-lee** [comment]
@zoechi I understand why they decided so. but.. If the reason for disabling the mirror system is only due to size, IMHO it's better option to let the each developer choose. 

if in-house application(user is specified), size is not a problem compare to usual cases
and we still need to use IoC Container library for enterprise application(or developers).
If possible, it would be a good idea to provide separate packages.

I also think the flutter team is still under [consideration](https://flutter.io/faq/#does-flutter-come-with-a-reflectionmirrors-system).

**Comment by jinseoung-lee** [comment]
Oh, There are also [dagger like compile-time ioc container library](https://github.com/google/inject.dart) released by Google.

**Comment by zoechi** [comment]
@jinseoung-lee it's not so easy to enable/disable such a feature. 
My understanding is that the `inject` package is in a very early stage and might not yet work.

**Comment by kidcosmic** [comment]
Just throwing this out there, it's immediately apparent `@MirrorsUsed` is a sufficient fix for an implementation for protecting tree shaking. Deliberate elimination of candidacy from tree shaking is not only an easy implementation, but sufficient in letting the shake flourish through the rest of the codebases. Mirror usage in the world is almost exclusively confined data/model classes which are never shaken as it is, injection which pretty much only has seemingly shakeable constructors and event bus implementations. These `MirrorsUsed`  cases only apply application side, such as in a widget and subscriptions to an event bus, but not package side, . The only exceptions there are monster application-like packages (ie have your data and render your ui). Incorrect `@MirrorsUsed` in packages are the only real threat to proper tree shaking, but that can and will likely be rooted out by the populace as well. Also mirrored data classes are by far better dev experience/speed than not, but that's debatable. Learning curves aren't usually an excuse.

Tldr; there is no _real_ good reason for no mirrors implementation.

That being said, I'm in the camp of who gives a shit. Codegen can get the job done. So screw it, codegen.

**Comment by Hixie** [comment]
If annotations are sufficient then you can also just implement a registration mechanism where in your main.dart you call a function that registers your class with your RTTI library (no code gen). Depends exactly what you want mirrors for, of course.

**Comment by zoechi** [comment]
@kidcosmic `@MirrorsUsed` wasn't satisfying for dart2js at all, this is why this direction was not pusued further.

**Comment by kidcosmic** [comment]
@Hixie ??? annotations = compile time, vs runtime. Annotations inform compiler? Funny thing is, I need to be able to the compiler "don't fuck with this code, i know better than you", since our compilers are not omniscient. If our compilers _were_ omnisicient_, then they would know what's being reflected and would be able to tree shake with reflections anyway. Annotations are a hack from point A to point B. A _legitimate_ implementation of mirroring with tree shaking support is of course our omniscient compilers. Also, where's my proguard.txt?

@zoechi never looked at their implementation. I was more basing the argument off of a signal compiler to remove candidacy implementation. Nuking candidacy is actually the quickest way to achieve mirrors + tree shaking  Mind you, it kind of looks like `@MirrorsUsed` kind of fell apart because of it's language. An implementation like proguards seems like it would flourish.

**Comment by Hixie** [comment]
@kidcosmic A registration mechanism does the same thing (because registering the identifier is a way to use it, and thus prevents it from being tree-shaken away).

**Comment by kidcosmic** [comment]
@Hixie ah i see, never had experience with RTTI libs, only recently started diving into native. From what I'm seeing in RTTI cpp examples (wikis) though is no registration mech in main? You saying that mirrors for example would need one? RTTI looks to be the same as a mirror/reflection lib O_o I guess you might be referring to `reflectabe`? Which i agree, does the job pretty damn well.

**Comment by Hixie** [comment]
RTTI is the same as mirrors or reflection or type inspection or any of the other ways to refer to things being discussed in this thread. It's just the more generic term.

I don't know exactly what you want mirrors for, but let's take the common case of wanting to be able to build instances of objects from a file on disk. A trivial runtime type information library for this purpose could look like:

```dart
// RTTI LIBRARY

typedef T Constructor<T>();

final Map<String, Constructor<Object>> _constructors = <String, Constructor<Object>>{};

void register<T>(Constructor<T> constructor) {
  _constructors[T.toString()] = constructor;
}

dynamic build(String type) {
  return _constructors[type]();
}
```

Here's how it could be used (here I've hard-coded the strings but they could come from disk or the network or whatever just as easily):

```dart
// THE VARIOUS CLASSES

class A { void test() => print('a'); }
class B { void test() => print('b'); }
class C { void test() => print('c'); }

void registerClasses() {
  // Register these classes with the RTTI library.
  // This is equivalent to putting an annotation on the classes.
  register<A>(() => new A());
  register<B>(() => new B());
  register<C>(() => new C());
}


// THE TEST
 
void test() {
  build('A').test(); // prints "a"
  build('B').test(); // prints "b"
  build('C').test(); // prints "c"
}


// THE APP

void main() {
  registerClasses();
  test();
}
```

**Comment by kidcosmic** [comment]
Ah I see, pretty clean with the constructors, so KV init would look like:

```dart
typedef T KVInit<T>(Map<String, dynamic> data);

dynamic build(String type, Map<String, dynamic> data);

class A { String name; }
class B { A a; }

void registerClasses() {
  register<A>((data) {
    var a = new A();
    a.name = data["name"];
    return a;
  });
  register<B>((data) {
    var b = new B();
    b.a = build('A', data['a']);
    return b;
  });
}
```

Not too shabby, make the classes pretty immediately available to db ORMs &  json IO.

**Comment by Hixie** [comment]
You can also put the constructors in the classes and have them be immutable:

```dart
class A { A(this.name); final String name; }
class B { B(this.a); final A a; }

void registerClasses() {
  register<A>((data) => new A(data['name']));
  register<B>((data) => new B(build('A', data['a'])));
}
```

**Comment by jerinho** [comment]
> I know that mirrors is blocked now because it causes the generated packages to be very large.
> 
> However, on the long term, it should remain a goal to find a solution for this. Many libraries use mirrors so not being able to use mirrors for Flutter apps will make the framework a lot less powerful.
> 
> The trade-off between large package sizes and using mirrors should be left to the user. A possible solution might be to use an explicit `--allow-mirrors` flag. But of course the ultimate solution should be to find a way to use mirrors without blowing up package sizes.
> 
> Java has reflection and Objective-C has Runtime Reference, so why would Flutter not have a reflection system?

where to put that `--allow-mirrors` ?

**Comment by zoechi** [comment]
@matejthetree as already mentioned above https://github.com/flutter/flutter/issues/1150#issuecomment-170231782 (and the following discussion)

**Comment by rrousselGit** [comment]
I don't think https://github.com/flutter/flutter/issues/26125 is a duplicate of this one, but I'll put my points here.

I personally agree that we do not need `dart:mirror` in a shipped application, as code-generation solves most of it.

But there's a catch: code-generators themselves may want to depend on flutter

A typical example is a package of mine: https://github.com/rrousselGit/functional_widget

It is a code generator that generates `Widget` subclass and `debugFillProperties` overrides. As such, `functional_widget` heavily depends on Flutter features.

But due to `dart:mirror` being disabled, it is impossible for such code generator to depend on Flutter as some generations dependencies use mirroring (such as `build_runner` and `source_gen`).

____


Overall, I think all dev dependencies should be allowed to use mirroring. The use-case is not limited to code generation, and it has no impact on what is shipped in the final app.

A typical use case is to add custom testing features similar to Jest. With mirroring, it becomes possible to use `dart2json` within tests and therefore make a `toMatchSnapshot`.

It may also be possible to use reflection for more advanced features, like running only tests impacted by the git local changes (like Jest again).

**Comment by rmawatson** [comment]
This should definitely be made available to depend on for dev_dependencies. It already works and is used by source_gen such as in TypeChecker.fromRuntime, but an import of `dart:mirrors` in a flutter project generator will be shown as invalid in IntelliJ IDEA and code completion will not work. 

I think if anything a prominent warning should be shown if a flutter project depends on mirrors, but it should be up to the user to be competent enough to manage their dependencies so as not to use mirrors in the actual app, if a bloated binary is the problem, rather then this being disabled altogether.

**Comment by paulcbetts** [comment]
I really wish this decision would be reconsidered, it's having some pretty significant effects across the ecosystem - having something as extremely common as JSON deserialization requiring people to manually set up build-time codegen is a *huge* barrier to adoption; no other mobile platform requires this.

Xamarin has a similar problem with AOT and Reflection, and solves it via allowing the developer to explicitly opt-in classes via linker directives - while it can be a pain in the ass in large apps to whack-a-mole types over and over until your app works, it is at least a Straightforward, Easy To Use process.

While alternatives such as `reflectable` exist, the fact that it can't be used in a library (i.e. you shouldn't publish `*.reflectable.dart` files) makes it pretty untenable as a General Solution to this problem

**Comment by rrousselGit** [comment]
> requiring people to manually set up build-time codegen is a huge barrier to adoption; no other mobile platform requires this.

Hopefully, Flutter did some progress on that topic. On master branch, code-gen integration is _working out of the box_ (at least that's what is promised). 

So this should make libraries that requires code-generators a lot more viable.

**Comment by eernstg** [comment]
@paulcbetts wrote:
> you shouldn't publish `*.reflectable.dart` files

This is a trade-off. Given that reflectable allows you to specify things like `@GlobalQuantifyCapability(r"^dart.core.List", reflector)`, which means that every subtype of `List` will support the reflective capabilities specified for `reflector`, there is no way you could generate code that would be able to fulfill that promise for subtypes of `List` declared in an arbitrary library of the "current program" unless you allow the code generator to run in the context of the client library. So you wouldn't publish `*.reflectable.dart` files, because they won't contain the code for the whole program, only for code reachable from your package.

But if you're happy saying that the classes for which there is reflection support is a closed set (that is, everything that you can see from your published package is covered, but you don't care that it won't work for declarations written in client packages), then you could publish your `*.reflectable.dart` files.

**Comment by ishowfun** [comment]
To flutter:越俎代庖

**Comment by Nastya-Scherbakova** [comment]
Well, that's really sad. Many things like json serialization/deserealization and SQLite creation of tables, requests are not comfortable to work with. I thought I can improve that using `dart:mirrors`, but has found this issue. It causes bad experience on projects with many classes.

**Comment by kasp1** [comment]
My reason to use mirrors was to set a user property in a user class from my library that doesn't know the property/object name in advance.

Please suggest a better solution without mirrors if you can think of any: https://stackoverflow.com/questions/58493417/assign-property-from-external-object/58501443#58501443

**Comment by matejthetree** [comment]
Check reflectable package. It provides mirrors with code generation.

**Comment by kasp1** [comment]
@matejthetree Quick check, does code generation require me or the user to use commands in their IDE that would generate additional files in the project/lib?

**Comment by matejthetree** [comment]
Not in the IDE, rather in terminal.

Check the package, it has rather extensive documentation.

**Comment by kasp1** [comment]
@matejthetree I personally believe simplicity of usage (and ease of learning) are the keys to success of any lib/framework/app. For this reason, I'd prefer to avoid code generation (as the user always has to learn something additional) and keep my "creative" syntax instead.

I'm not blaming Flutter, it's a great framework and if disabling mirrors is a worthy trade off for something else, fine then. Maybe as it's been already mentioned here, optional mirrors support might be handy but I can understand it may be alot of work. Just adding here a reason to do so.

**Comment by jmendeth** [comment]
So far I haven't seen a technical explanation of what it takes to allow the user to use dart:mirrors... Isn't tree shaking performed by the snapshotter, i.e. Dart? Why/how is Flutter involved in this decision?

**Comment by icatalud** [comment]
Quantifying the cost of adding mirrors should be clarified, so people can get convinced it's not good idea to add mirrors. Say something like "This is what would happen to the program if mirrors is enabled: uses as 200mb" (for example).

Could someone from the Dart team give an idea of the current cost?

**Comment by jmendeth** [comment]
I'm not a Dart expert but I can now confirm *this has nothing to do with Flutter*, it's entirely a Dart issue, and I'll try to give some technical insights.

There are actually two places in the AOT compilation pipeline where tree shaking is performed:

1. In the frontend server, when compiling Dart code to the kernel binary (`.dill`): Type Flow Analysis is performed ([source code](https://github.com/dart-lang/sdk/blob/cb6127570889bed147cbe6292cb2c0ba35271d58/pkg/vm/lib/transformations/type_flow/transformer.dart)), and a two-pass tree shaking is done as part of this process.

2. In `gen_snapshot`, which compiles the kernel binary to an AOT snapshot (`libapp.so`): in the precompilation step ([`Dart_Precompile`](https://github.com/dart-lang/sdk/blob/cb6127570889bed147cbe6292cb2c0ba35271d58/runtime/include/dart_api.h#L3178) which calls [`Precompiler::DoCompileAll`](https://github.com/dart-lang/sdk/blob/cb6127570889bed147cbe6292cb2c0ba35271d58/runtime/vm/compiler/aot/precompiler.cc#L190)) a second tree shaking is performed.

In my experiments the second tree-shaking barely removed anything that the TFA hadn't done already.

I've patched out the tree-shaking entirely: in a test project, it increased the snapshot file size from 9MB to 30MB and snapshot loading times by up to 5x. So yeah, it's quite an important optimization. I haven't made any RAM / performance tests.

However I think it'd be easy to process `@MirrorsUsed` annotations in the tree shaker and skip those, just as it's already done with the `vm:entry-point` pragma. I still have a lot to learn about the VM, so I could be missing some important details...

**Comment by k-paxian** [comment]
If someone is looking for a decent solution based on ```reflectable``` I've got it
https://github.com/k-paxian/dart-json-mapper
@eernstg Thank you for all your hard work on that approach.

**Comment by pedromassango** [comment]
> Overall, I think all dev dependencies should be allowed to use mirroring. The use-case is not limited to code generation, and it has no impact on what is shipped in the final app.
> 
I agree with @rrousselGit. Most of the libraries that uses Code generator (and want to use dart:mirrors) are dev dependencies so maybe they don't impact the binary size.

**Comment by aspectgfg** [comment]
Pre-mirrors support here:
While I think eliminating mirror support is awful because dev's carry the cost of mirroring and are fully capable of handling mirroring themselves, the generator libs are actually pretty awesome and satisfy needs of developers very well.
